metagraph/src/graph/representation/canonical_dbg.hpp

```diff
@@ -112,6 +141,28 @@ class CanonicalDBG : public DBGWrapper<DeBruijnGraph> {
     bool has_sentinel_;
 
     std::array<size_t, 256> alphabet_encoder_;
+
+    // a thread-safe 0-size NodeFirstCache to use as a fallback if another NodeFirstCache
+    // extension is not available
+    mutable std::unique_ptr<NodeFirstCache> fallback_cache_;
+
+    // Find incoming nodes that are on the reverse complement strand of node
+    void adjacent_incoming_rc_strand(node_index node,
+                                     const std::function<void(node_index)> &callback,
+                                     const std::string &spelling_hint) const;
+    void call_incoming_rc_strand(node_index node,
+                                 const std::function<void(node_index, char)> &callback,
+                                 const std::string &spelling_hint) const;
+
+    // Find outgoing nodes that are on the reverse complement strand of node
+    void adjacent_outgoing_rc_strand(node_index node,
+                                     const std::function<void(node_index, uint64_t)> &callback,
+                                     const std::string &spelling_hint) const;
+    void call_outgoing_rc_strand(node_index node,
+                                 const std::function<void(node_index, char)> &callback,
+                                 const std::string &spelling_hint) const;
```

Comment:
There are so many of these functions... Do we really need all of them?

Reply:
Having a separate `adjacent_outgoing_nodes` (which the extra functions are meant to support) is used by #453 to find the descendants of dummy k-mers when annotating dummy k-mer seeds.